azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java

```diff
@@ -71,24 +77,88 @@
+        final WebAppConfig config = getWebAppConfig();
+        final IAppService target = createOrUpdateResource(config);
+        deployArtifacts(target, config);
+    }
+
+    private IAppService createOrUpdateResource(final WebAppConfig config) throws AzureExecutionException {
+        final IAppService target = getDeployTarget(config);
+        if (target instanceof IWebApp) {
+            return target.exists() ? updateWebApp((IWebApp) target, config) : createWebApp((IWebApp) target, config);
+        } else {
+            return target.exists() ? updateDeploymentSLot((IWebAppDeploymentSlot) target, config) :
+                    createDeploymentSlot((IWebAppDeploymentSlot) target, config);
+        }
+    }
+
+    private IAppService getDeployTarget(final WebAppConfig config) throws AzureExecutionException {
+        final IWebApp webApp = getAppServiceClient().webapp(config.getResourceGroup(), config.getAppName());
+        final boolean isDeploymentSlot = StringUtils.isNotEmpty(config.getDeploymentSlotName());
+        if (isDeploymentSlot && !webApp.exists()) {
+            throw new AzureExecutionException(WEBAPP_NOT_EXIST_FOR_SLOT);
+        }
+        return isDeploymentSlot ? webApp.deploymentSlot(config.getDeploymentSlotName()) : webApp;
+    }
+
+    private IWebApp createWebApp(final IWebApp webApp, final WebAppConfig webAppConfig) {
+        // todo: get or create Resource Group
+        final ResourceGroup resourceGroup = null;
+        // Get or create App Service Plan
+        final IAppServicePlan appServicePlan = getAppServiceClient()
+                .appServicePlan(webAppConfig.getServicePlanResourceGroup(), webAppConfig.getServicePlanName());
+        if (appServicePlan.exists()) {
+            appServicePlan.create()
+                    .withName(webAppConfig.getServicePlanName())
+                    .withResourceGroup(resourceGroup.getName())
+                    .withRegion(webAppConfig.getRegion())
+                    .withPricingTier(webAppConfig.getPricingTier())
+                    .withOperatingSystem(webAppConfig.getRuntime().getOperatingSystem())
+                    .commit();
+        }
+        return webApp.create().withName(webAppConfig.getAppName())
+                .withResourceGroup(webAppConfig.getResourceGroup())
+                .withPlan(appServicePlan.entity().getId())
+                .withRuntime(webAppConfig.getRuntime())
+                .withDockerConfiguration(webAppConfig.getDockerConfiguration())
+                .withSubscription(webAppConfig.getSubscriptionId())
+                .withAppSettings(webAppConfig.getAppSettings())
+                .commit();
+    }
+
+    private IWebApp updateWebApp(final IWebApp webApp, final WebAppConfig webAppConfig) throws AzureExecutionException {
+        // update app service plan
+        final IAppServicePlan currentPlan = getAppServiceClient().appServicePlan(webApp.entity().getAppServicePlanId());
```

Comment:
it's weird to create a `AppServicePlan instance from `WebAppEntity`
I would suggest
```
webapp.appServicePlan().entity().getId();
```